cmake/grpc_generate_cpp.cmake

```
@@ -0,0 +1,104 @@
+# SPDX-FileCopyrightText: Copyright (c) 2024 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+# SPDX-License-Identifier: Apache-2.0
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# A helper function to generate C++ code from given .proto files.
+cmake_minimum_required(VERSION 3.8)
+find_package(Threads REQUIRED)
+
+function(grpc_generate_cpp SRCS HDRS INCLUDE_DIRS)
+    # Expect:
+    # - PROTOC_EXECUTABLE: path to protoc
+    # - GRPC_CPP_EXECUTABLE: path to grpc_cpp_plugin
+    if(NOT ARGN)
+        message(SEND_ERROR "Error: grpc_generate_cpp() called without any .proto files")
+        return()
+    endif()
+
+    foreach(PROTO_FILE ${ARGN})
+        message(STATUS "Build proto file ${PROTO_FILE}")
+        # Get the full path to the proto file
+        get_filename_component(_abs_proto_file "${PROTO_FILE}" ABSOLUTE)
+        # Get the name of the proto file without extension
+        get_filename_component(_proto_name_we ${PROTO_FILE} NAME_WE)
+        # Get the parent directory of the proto file
+        get_filename_component(_proto_parent_dir ${_abs_proto_file} DIRECTORY)
+        # Get the parent directory of the parent directory
+        get_filename_component(_parent_dir ${_proto_parent_dir} DIRECTORY)
+        # Append 'generated' to the parent directory
+        set(_generated_dir "${_parent_dir}/generated")
```

Could we move output to the binary directory to avoid committing generated sources in git history?

```suggestion
        set(_generated_dir "${CMAKE_CURRENT_BINARY_DIR}/generated")
```

---

A `.gitignore` file was added to avoid that.